### ISSUE TITLE
shorten KBlock propose/validate time

### ISSUE THREAD

**Issue by simonzg** [meterio/meter-pov#18]
Need a thorough investigation into this.

The KBlock usually takes more than 10s to build, this already adds more delay on proposer. And on validator side, current validation scheme needs to propose a local KBlock and compare each and every tx. So from the network perspective, it requires more than double the proposed time in order to let everyone on the committee agree on a specific KBlock, and this usually is more than 25s, which is way too much compared to our avgBlockTime of 2.4s.

My first guess is `autobid` tx takes much longer than expected, since it has a lot of clauses. Writing a test to build KBlock from a specific height might be helpful to pinpoint the root cause.

**Comment by simonzg** [comment]
Could do two improvements:

1. merge these txs together into one epoch "clearing" tx: staking governing, account lock governing, autobid
2. in order to make sure everyone has the same result, the `timestamp` used must be over consensus, and we should probably use timestamp from parent block instead of proposing time. It will introduce acceptable error. And with this, everyone should be able to re-calc everything by themselves with the same timestamp and tx don't need to carry any input data anymore.

**Comment by simonzg** [comment]
Time for proposing kblock includes 2 parts:

1. build kblock txs
2. adopt kblock txs (executing each and every txs to get state root)

My observation is that kblocks right after the new auction start is usually fast and stable, and kblocks before the auction close takes much longer to build. My first guess is the auction txs are accumulating during each auction and handling bid towards the end is much slower. Haven't figured out the root cause, just added logs to print out elapsed time for each steps and we'll see.

**Comment by simonzg** [comment]
![Screen Shot 2023-01-05 at 2 22 15 PM](https://user-images.githubusercontent.com/7259860/210891492-bc77b60b-ef80-4173-ab5b-7b6d3e7cb7fb.png)

The main issue, as expected, was caused by adopting autobid tx and auction close tx

Will continue to look into either IO or sort is causing the issue in both cases

**Comment by simonzg** [comment]
narrowed down with logs

```
t=2023-01-06T03:47:56+0000 lvl=info msg="Read completed. " pkg=auction getAuction=112ns getState=49ns getAuctionCB=7.646ms
t=2023-01-06T03:47:56+0000 lvl=info msg="Precheck completed" pkg=auction elapsed=7.731Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="New Auction Tx created" pkg=auction elapsed=15.705Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="Get completed" pkg=auction elapsed=2.217Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="Add MTR completed" pkg=auction elapsed=1.348Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="indexOf completed" pkg=auction elapsed=765ns
t=2023-01-06T03:47:56+0000 lvl=info msg="insert completed" pkg=auction elapsed=54.203Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="Add completed" pkg=auction elapsed=109.245Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="Auction tx added" pkg=auction elapsed=188.367Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="Transfer funds completed" pkg=auction elapsed=9.239Âµs
t=2023-01-06T03:47:56+0000 lvl=info msg="Save completed" pkg=auction elapsed=3.261ms
t=2023-01-06T03:47:56+0000 lvl=info msg="Bid completed" pkg=auction elapsed=11.530ms
```

you can see for one bid, it takes `11.53ms`, and io (read/save) takes `10.97ms`, which is over 94.5%

**Comment by simonzg** [comment]
Did some digging and noticed, RLP encode/decode is the real bottleneck. I've tried to serialize/deserialize AuctionCB with a test case and here's the log, it indicates rlp encode/decode time is proportional to number of txs in AuctionCB.

```
Epoch 1 , Auction Txs:  659
RLP decode: 557.859µs , encode: 492.162µs
Epoch 2 , Auction Txs:  1318
RLP decode: 2.051ms , encode: 586.71µs
Epoch 3 , Auction Txs:  1977
RLP decode: 1.370ms , encode: 738.387µs
Epoch 4 , Auction Txs:  2636
RLP decode: 2.262ms , encode: 1.126ms
Epoch 5 , Auction Txs:  3295
RLP decode: 2.105ms , encode: 1.609ms
Epoch 6 , Auction Txs:  3954
RLP decode: 2.890ms , encode: 1.056ms
Epoch 7 , Auction Txs:  4613
RLP decode: 2.831ms , encode: 1.870ms
Epoch 8 , Auction Txs:  5272
RLP decode: 3.388ms , encode: 2.193ms
Epoch 9 , Auction Txs:  5931
RLP decode: 3.520ms , encode: 1.898ms
Epoch 10 , Auction Txs:  6590
RLP decode: 4.577ms , encode: 2.728ms
Epoch 11 , Auction Txs:  7249
RLP decode: 4.211ms , encode: 1.630ms
Epoch 12 , Auction Txs:  7908
RLP decode: 4.455ms , encode: 2.463ms
Epoch 13 , Auction Txs:  8567
RLP decode: 5.449ms , encode: 1.958ms
Epoch 14 , Auction Txs:  9226
RLP decode: 5.743ms , encode: 2.007ms
Epoch 15 , Auction Txs:  9885
RLP decode: 6.241ms , encode: 4.701ms
Epoch 16 , Auction Txs:  10544
RLP decode: 6.102ms , encode: 3.154ms
Epoch 17 , Auction Txs:  11203
RLP decode: 6.492ms , encode: 3.000ms
Epoch 18 , Auction Txs:  11862
RLP decode: 7.270ms , encode: 4.980ms
Epoch 19 , Auction Txs:  12521
RLP decode: 7.624ms , encode: 2.715ms
Epoch 20 , Auction Txs:  13180
RLP decode: 7.489ms , encode: 3.153ms
Epoch 21 , Auction Txs:  13839
RLP decode: 8.047ms , encode: 3.432ms
Epoch 22 , Auction Txs:  14498
RLP decode: 8.531ms , encode: 3.120ms
Epoch 23 , Auction Txs:  15157
RLP decode: 8.839ms , encode: 3.599ms
```

**Comment by simonzg** [comment]
So now the solution would be: merge autobid clauses into staking governing, and this requires:

1. Use `RewardInfoV2` to serialize for `ExtraData` in staking governing StakingBody
2. Handle autobids in staking governing handler
3. This will be enabled after fork 6

**Comment by simonzg** [comment]
implemented in 4e260f88cbfe213b3856d4c06fff2c5fe9d2fe50 and 5e86c3971dea60ded169a67e25bcc12cadc89498 and
